bool dont_clear_first